Edits for README.adoc (#18)
OpenLiberty/guide-grpc-intro: README.adoc

:common-includes: https://raw.githubusercontent.com/OpenLiberty/guides-common/prod
:imagesdir: /img/guide/{projectid}
:source-highlighter: prettify
:page-seo-title: Streaming messages between client and server services using gRPC remote procedure calls
:page-seo-description: A getting started tutorial with examples on how to write gRPC remote procedure call services with unary calls, server streaming, client streaming, and bidirectional streaming using Java with Open Liberty.
:guide-author: Open Liberty
= Streaming messages between client and server services using gRPC

[.hidden]
NOTE: This repository contains the guide documentation source. To view the guide in published form, view it on the https://openliberty.io/guides/{projectid}.html[Open Liberty website].

Learn how to use gRPC unary calls, server streaming, client streaming, and bidirectional streaming to communicate between Java client and server services with Open Liberty.

// ===============================================================================================
//  What you'll learn
// ===============================================================================================

== What is gRPC?

> The https://grpc.io/[gRPC^] Remote Procedure Call is a technology that implements remote procedure call (RPC) style APIs with HTTP/2. Typically, gRPC uses https://developers.google.com/protocol-buffers/docs/reference/overview[protocol buffers^] to define format of data to be transferred and service interfaces to access it, which include service calls and expected messages. For each service defined in a `.proto` file, gRPC uses the definition to generate the skeleton code for users to implement and extend. Protocol buffers use a binary format to send and receive messages that is faster and more lightweight than the JSON that is typically used in RESTful APIs.

Reviewer: ```suggestion
The https://grpc.io/[gRPC^] Remote Procedure Call is a technology that implements remote procedure call (RPC) style APIs with HTTP/2. Typically, gRPC uses https://developers.google.com/protocol-buffers/docs/reference/overview[protocol buffers^] to define the format of data to be transferred and the service interfaces to access it, which include service calls and expected messages. For each service defined in a `.proto` file, gRPC uses the definition to generate the skeleton code for users to implement and extend. Protocol buffers use a binary format to send and receive messages that is faster and more lightweight than the JSON that is typically used in RESTful APIs.
```

> Protocol buffers allow cross-project support through the `.proto` file. As a result, gRPC clients and servers can run and communicate with each other from different environments. For example, a gRPC client running on a Java virtual machine is able to call a gRPC server developed in other https://grpc.io/docs/languages/[supported languages^]. This feature of protocol buffers allows for easier integration between services.

Reviewer: ```suggestion
Protocol buffers allow cross-project support through the `.proto` file. As a result, gRPC clients and servers can run and communicate with each other from different environments. For example, a gRPC client running on a Java virtual machine can call a gRPC server developed in any other https://grpc.io/docs/languages/[supported language^]. This feature of protocol buffers allows for easier integration between services.
```

== What you'll learn

You will learn how to create gRPC services and their clients by using protocol buffers and how to implement them with Open Liberty. You will use Maven to generate the gRPC stubs, deploy the services, and to interact with the running Liberty runtime.

The application that you will build in this guide consists of three projects: the `systemproto` model project, the `query` client service, and the `system` server service.

The `query` service implements four RESTful APIs by using four different gRPC streaming methods.

* Unary RPC: The client sends a single request and receives a single response.
* Server streaming RPC: The client sends a single request and the server returns a stream of messages.
* Client streaming RPC: The client sends a stream of messages and the server responds with a single message.
* Bidirectional RPC: Both client and server send a stream of messages. The client and server can read and write messages in any order.

> image::architecture.png[Application architecture of the gRPC application covered in guide where a system service acts as the gRPC server and a query service acts a client that makes different gRPC streaming calls, align="center"]

Reviewer: ```suggestion
image::architecture.png[Application architecture of the gRPC application covered in guide, where a system service acts as the gRPC server and a query service acts a client that makes different gRPC streaming calls, align="center"]
```

----
include::finish/query/src/test/java/it/io/openliberty/guides/query/QueryIT.java[]
----

> Each test case tests one type of the gRPC calls that have been implemented.

Reviewer: ```suggestion
Each test case tests one type of the gRPC calls that you implemented.
```
